architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+
+When the GitOps tool watches the repository and applies the BicepDeployment resource to the production cluster, the Radius resources defined in the `app.bicep` Bicep manifest are deployed.
+
+#### Sam can use GitOps to patch a Radius application in production
+
+Sam is an SRE who is responsible for maintaining a Radius application in production. He notices that the application is experiencing performance issues and the number of application container replicas needs to be increased. He updates the Bicep parameters file for the application to specify the new number of replicas and runs the `rad bicep generate bicepdeployment app.bicep --parameters app.bicepparam` command on his machine to create an BicepDeployment resource that contains the updated parameters. He then commits the BicepDeployment resource to the Git repository.
+
+When the GitOps tool watches the repository and applies the BicepDeployment resource to the production cluster, the Radius resources defined in the `app.bicep` Bicep manifest are updated with the new parameters. The application container replicas are increased, and the performance issues are resolved.
+
+
+## User Experience
+<!--
+If the change impacts the user experience, provide expected interaction 
+flow we aim to achieve through this proposal.
+
+When users interact with Radius through the CLI, include sample 
+input commands and their corresponding output. Include a bicep/helm code 
+sample, if this proposal involves updates to that experience.
+-->
+
+**Sample Input:**
+<!--
+Provide a sample CLI command input and/or bicep/helm code.
+-->
+
+#### `env.bicep`
+```bicep
+extension radius
+
+param k8sNamespace string
+
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'env'
+  properties: {
+    compute: {
+      kind: 'kubernetes'
+      namespace: k8sNamespace
+    }
+  }
+}
+```
+
+#### `env.bicepparam`
+```bicep
+using './env.bicep'
+
+param k8sNamespace string = 'demo'
+```
+
+**Sample Output:**
+<!--
+Provide a sample output for the inputs provided above.
+-->
+```
+> rad bicep generate bicepdeployment env.bicep --parameters env.bicepparam --out env.yaml
+
+Generating BicepDeployment resource...
+BicepDeployment resource generated at env.yaml
+
+To apply the BicepDeployment resource onto your cluster, run:
+kubectl apply -f env.yaml
+```
+
+#### `env.yaml`
+```yaml
+kind: BicepDeployment
+apiVersion: radapp.io/v1alpha3
+metadata:
+  name: env
+  namespace: radius-system
+spec:
+  template: |
+    extension radius
+
+    param k8sNamespace string
+
+    resource env 'Applications.Core/environments@2023-10-01-preview' = {
+      name: 'env'
+      properties: {
+        compute: {
+          kind: 'kubernetes'
+          namespace: k8sNamespace
+        }
+      }
+    }
+  parameters: |
+    using './env.bicep'
+
+    param k8sNamespace string = 'demo'
+```
```

Comment:
Do we need more fields in the initial CR?

Reply:
Based on this design we need:

- Everything required to compile bicep:
   - Bicep file
   - Modules referenced as local files
   - `bicepconfig.json`
   - Do we need to know the version of the compiler?

- Everything required to execute bicep:
   - (compiler output)
   - parameter files? (can there be multiple?)
   - does it make sense to support additional non-file-based parameters?
   - configuration of `extensions`
       - eg: azure scope
       - eg: AWS scope
       - eg: Radius scope + application + environment

We might want to talk to the Bicep team about this idea sooner rather than later.